Better mod blacklisting

Tiny-DM opened the pull request:
## What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [X] Enhancement

### What have you changed and why?
Provides a better blacklisting system for mods
Each banned mod will put a hintC on screen that needs to be dismissed.
Put entries for each banned mod in the guide. Also banned ACE Medical in SP

### Please verify the following and ensure all checks are completed.

1. [X] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [X] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: Could throw it on dedicated and make sure the server side detection works, which isnt a too-too critical feature because hopefully either the server is a player who gets the errors or its a dedicated which is enforcing signatures, but it's still useful to have

********************************************************
Notes:

Tiny-DM committed:
setup new mod blacklist handling

Tiny-DM committed:
add ace medical block in sp

jaj22 commented on A3A/addons/core/functions/init/fn_modBlacklist.sqf:
```diff
@@ -1,23 +1,55 @@
-_bad = false;
+if !(canSuspend) exitWith {};
+
+private _blacklistedHM = createHashMapFromArray [ // ["Mod Name As Seen On Steam", [{condition}, "Dev Comment"],
+	// All of these strings dont need to be localized
+	["LAMBS_Danger", [{isClass (configfile >> "CfgPatches" >> "lambs_danger")}, "Causes AI to walk away from garrisons and ignore tasks assigned by Antistasi"]],
+	["Vcom", [{isClass (configfile >> "CfgPatches" >> "VCOM_AI")}, "Breaks Antistasi AI and also causes AI to attempt to steal random vehicles"]],
+	["ALiVE", [{isClass (configfile >> "CfgVehicles" >> "ALiVE_require")}, "Antistasi has its own save system and AI command system"]],
+	["MCC Sandbox 4", [{missionNamespace getVariable ["MCC_isMode",false]}, ""]], // blocked from a while ago, could find no solid info
+	["ASR AI3", [{"asr_ai3_main" in activatedAddons}, ""]], // blocked from a while ago, could find no solid info
+	["Project Injury Reaction (PiR)", [{isClass (configfile >> "CfgPatches" >> "PiR")}, "Breaks the death / unconscious states for Antistasi medical. May work fine with ACE Medical."]],
+	["Zulu Headless Client (ZHC)", [{isClass (configfile >> "CfgPatches" >> "zhc_main")}, "Antistasi is reliant on its own scripts for using headless clients"]],
+	["Werthles' Headless Module", [{isClass (configfile >> "CfgPatches" >> "Werthles_WHK")}, "Antistasi is reliant on its own scripts for using headless clients"]],
+	["Advanced Rappelling", [{isClass (configfile >> "CfgPatches" >> "AR_AdvancedRappelling") && (isServer && (productVersion # 6) isEqualTo "Linux")}, "Breaks AI movement on Linux servers"]],
+	["Advanced Urban Rappelling", [{isClass (configfile >> "CfgPatches" >> "AUR_AdvancedUrbanRappelling") && (isServer && (productVersion # 6) isEqualTo "Linux")}, "Breaks AI movement on Linux servers"]],
+	["ACE Medical", [{isClass (configFile >> "CfgSounds" >> "ACE_heartbeat_fast_3") && {!isMultiplayer}}, "Breaks respawning in singleplayer. Download the ACE No Medical mod to get the rest of the ACE features without the broken Medical"]], // broken only in real sp
+];
```
Hmm. I feel like these should be passive CfgPatches entries with an optional extra condition. But I guess this works.

Tiny-DM replied:
Yeah I just pulled from the existing system for those conditions

jaj22 commented on A3A/addons/core/functions/init/fn_modBlacklist.sqf:
```diff
@@ -1,23 +1,55 @@
+		_unit setVariable ["lambs_danger_disableAI", true];
+	}] call CBA_fnc_addClassEventHandler;
+};
+
+if !(hasInterface) exitWith {
+	waitUntil {sleep 0.5; count playableUnits > 0};
+	[localize "STR_A3A_mod_blacklist_title", format [localize "STR_A3A_mod_blacklist_serverHint", _badMods joinString "<br/>"]] call A3A_fnc_customHint;
 };
```
If it doesn't have an interface then customHint isn't gonna work on it. Could maybe wait until there's an admin and hint it to them instead.

Tiny-DM replied:
Yeah I think the intent here was to remoteExec onto whoever the current admin was when there was at least one player (so theres guaranteed to be an admin). Think I forgot that part though and made it a local call

jaj22 replied:
Existence of players doesn't guarantee an admin anyway, even at that point. autoLoadLastGame is real.

Tiny-DM replied:
Even assigning it to the first player should be fine; they'll be autoloading the game in that case. As long as someone ends up seeing it. This also isnt critical, the only real use case is a dedicated server with signature verification off, which is enough of an edge case.

jaj22 replied:
Still doesn't do anything btw.

jaj22 commented on A3A/addons/core/functions/init/fn_initClient.sqf:
```diff
@@ -26,7 +26,7 @@ if (!isNil "serverInitDone" and !isNil "A3A_utilityItemHM") then {
 
 if (!requiredVersion QUOTE(REQUIRED_VERSION)) exitWith { Error("Arma version is out of date") };
-if (call A3A_fnc_modBlacklist) exitWith {};
+[] spawn A3A_fnc_modBlacklist;
```
Yeah the trouble here is that there's no sequencing control and therefore this fights with other UIs such as the setup UI and the version mismatch customHint. I can't say what that's going to look like without testing the cases.

Tiny-DM replied:
It goes on top of other UIs and needs to be manually dismissed, so I dont think its much of an issue. It does probably overwrite the version mismatch hint but I dont think theres a reasonable fix for that, we were gonna look into changing the version mismatch feedback anyways, and if you're running both bad mods and the wrong version then it's just fucked

jaj22 commented on A3A/addons/core/functions/init/fn_modBlacklist.sqf:
```diff
@@ -1,23 +1,55 @@
 
-_bad;
+{
+	private _mod = _x;
+	private _devComment = (_blacklistedHM get _mod)#1;
+	if (_devComment isEqualTo "") then {_devComment = "None"};
+	sleep 0.1;
+	localize "STR_A3A_mod_blacklist_warningTitle" hintC parseText format [localize "STR_A3A_mod_blacklist_warningText", _mod, _devComment];
+	waitUntil {sleep 0.1; isNull findDisplay 72};
+	hintSilent "";
+} forEach _badMods;
```
Doesn't this section double-fire on localhost? Not sure what that looks like either.

jaj22 replied:
Also the LAMBS disable hack installs the EH twice I think? Should probably just not run this thing from the server if hasInterface is true.

Tiny-DM committed:
update 25.11.25

jaj22 commented on A3A/addons/core/functions/init/fn_modBlacklist.sqf:
```diff
@@ -1,6 +1,7 @@
 #include "..\..\script_component.hpp"
 FIX_LINE_NUMBERS()
 if !(canSuspend) exitWith {};
+if !(isNil "A3A_modsChecked") exitWith {};
```
Ew. Could have just put a hasInterface check on the initServer one.

jaj22 commented:
Remaining issues:
- Server->client feedback still doesn't do anything.
- Mod blacklist popup nukes the version mismatch message. Could make the client-side modBlacklist check a call (probably after background init), and then convert the version mismatch error to a hintC as well.

Tiny-DM committed:
update 2 25.11.25

jaj22 committed:
Mod blacklisting improvements:
- Improved display reliability & sequencing of hintCs.
- Server now sets public var with bad mods, client displays with hintC.
- Converted modBlacklist to 4-space.
- Semi-handled some headless client cases.

jaj22 commented:
Changes made after testing:
- Improved display reliability & sequencing of hintCs.
- Server now sets public var with bad mods, client displays with hintC.
- Converted modBlacklist to 4-space.
- Semi-handled some headless client cases.

I think it's solid enough now. Cases tested:
- Localhost: Clean, single & multiple bad mods, Altis & Tanoa.
- DS + client: Bad mods on one or both, version mismatch + bad mods, setup & JIP, Altis & Tanoa.
- DS + HC + client: Clean only.

jaj22 reviewed (approved):
Yeah technically I have to approve this myself.
